engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -25,29 +28,48 @@
  */
 public final class RenderTaskListGenerator {
     private static final Logger logger = LoggerFactory.getLogger(RenderTaskListGenerator.class);
+    private List<Object> intermediateList;
+    private List<RenderPipelineTask> taskList;
 
     public RenderTaskListGenerator() {
+        taskList = Lists.newArrayList();
+        intermediateList = Lists.newArrayList();
     }
 
     public List<RenderPipelineTask> generateFrom(List<Node> orderedNodes) {
-        logger.info("Task List: "); // TODO: remove in the future or turn it into debug
-        List<Object> intermediateList = generateIntermediateList(orderedNodes);
-        return generateTaskList(intermediateList);
+        generateIntermediateList(orderedNodes);
+        logList("-- Intermediate List --", intermediateList); // TODO: remove in the future or turn it into debug
+        generateTaskList();
+        logList("-- Task List --", taskList); // TODO: remove in the future or turn it into debug
+        return taskList;
     }
 
-    private List<RenderPipelineTask> generateTaskList(List<Object> intermediateList) {
-        List<RenderPipelineTask> taskList = Lists.newArrayList();
-
-        for (Object o : intermediateList) {
-            taskList.add(generateTask(o));
+    private void logList(String title, List<?> list) {
+        logger.info(title);
+        for (Object o : list) {
+            logger.info(o.toString());
         }
-
-        return taskList;
     }
 
-    private List<Object> generateIntermediateList(List<Node> orderedNodes) {
-        List<Object> intermediateList = Lists.newArrayList();
+    private void generateTaskList() {
+        taskList.clear();
+
+        Map map = Maps.newHashMap();
+        for (Object object : intermediateList) {
+            if (object instanceof StateChange) {
+                map.put(object.getClass(), object);
+            } else {
+                for (Object o : map.values()) {
+                    taskList.add(((StateChange) o).generateTask());
```

Comment:
In line with what mentioned below, you need to check first if a state change in the persistingStateChanges map makes the current state change redundant.